ports-of-call/robust_utils.hpp

```diff
@@ -64,13 +64,40 @@ PORTABLE_FORCEINLINE_FUNCTION auto ratio(const A &a, const B &b) {
   return a / (b + sgn(b) * SMALL<B>());
 }
 
+template <typename DataType>
+PORTABLE_FUNCTION constexpr DataType reldiff(const DataType actual, const DataType expected)
+{
+    if (actual == expected) {
+        return 0;
+    } else if (expected == 0) {
+        // expected 0, got nonzero, define this to be 100% error
+        return 1;
+    } else {
+        // Note: we do not use std::abs to calculate the absolute value because
+        //       it is not constexpr until C++23, therefore it is not portable
+        //       to GPUs (currently using C++17).
+        const auto signed_reldiff = (actual - expected) / expected;
+        return (signed_reldiff >= 0 ? signed_reldiff : -signed_reldiff);
+    }
+}
+
 template <typename T>
 PORTABLE_FORCEINLINE_FUNCTION T safe_arg_exp(const T &x) {
   return x < min_exp_arg<T>()   ? 0.0
          : x > max_exp_arg<T>() ? std::numeric_limits<T>::infinity()
                                 : std::exp(x);
 }
 
+// type-safe check against 0
+template <typename T, std::enable_if_t<std::is_unsigned<T>::value, bool> = true>
+PORTABLE_FUNCTION constexpr bool check_nonnegative(const T) {
+  return true;
+}
+template <typename T, std::enable_if_t<!std::is_unsigned<T>::value, bool> = true>
+PORTABLE_FUNCTION constexpr bool check_nonnegative(const T t) {
+  return t >= 0;
+}
+
```

Comment:
@Yurlungur and I discussed moving this out of `span.hpp`, because Nautilus and Singe both want to use it and I don't want to carry multiple implementations.